### PR TITLE
Feature: Get Folder Content

### PR DESCRIPTION
**Pull request by bohdan-meltiukhov**
# Get Folder Content
This PR adds the get folder request and allows navigating between folders.

### CODE REVIEW

File: ui-components/app/components/file-list-page/index.js [hunk 1]
```diff
@@ -16,12 +17,14 @@ export default class FileListPage extends StateAwareComponent {
    *
    * @param {Element} container - The parent element for the page.
    * @param {StateManager} stateManager - The state manager to use.
+   * @param {object} properties - The URL properties.
    */
-  constructor(container, stateManager) {
+  constructor(container, stateManager, properties) {
```
[the remark this answers is not in the document]
Comment (bohdan-meltiukhov) [reply ui-components/app/components/file-list-page/index.js:25]: I added a separate model for the URL properties.

File: ui-components/app/components/file-list-page/index.js [hunk 2]
```diff
@@ -16,12 +17,14 @@ export default class FileListPage extends StateAwareComponent {
...
     super(container, stateManager);
 
     this.render();
-    stateManager.dispatch(new GetFilesAction());
+    stateManager.dispatch(new GetFolderAction(properties.id));
```
[the remark this answers is not in the document]
Comment (bohdan-meltiukhov) [reply ui-components/app/components/file-list-page/index.js]: Done

File: ui-components/app/components/file-list-page/index.js [hunk 3]
```diff
@@ -100,24 +109,54 @@ export default class FileListPage extends StateAwareComponent {
...
-        this.stateManager.dispatch(new GetFolderAction(state.locationParameters.id));
-        this.stateManager.dispatch(new GetFilesAction(state.locationParameters.id));
+      if (state.locationParameters.folderId) {
+        this.stateManager.dispatch(new GetFolderAction(state.locationParameters.folderId));
+        this.stateManager.dispatch(new GetFilesAction(state.locationParameters.folderId));
       }
     });
 
+    this.onStateChanged('isFolderLoading', (event) => {
+      const state = event.detail.state;
```
Comment (dkalinin) [inline ui-components/app/components/file-list-page/index.js:130]:
```suggestion
    this.onStateChanged('isFolderLoading', ({detail: {state}}) => {
```

File: ui-components/app/components/login-page/index.js [hunk 4]
```diff
@@ -52,8 +53,7 @@ export default class LoginPage extends Component {
     const apiService = ApiService.getInstance();
     apiService.logIn(userCredentials)
       .then(() => {
-        const rootFolderRoute = FILE_LIST_ROUTE.slice(0, FILE_LIST_ROUTE.indexOf(':id')) + '4Goz0J0Tz8xfDfsJ';
-        window.location.hash = rootFolderRoute;
+        window.location.hash = FILE_LIST_ROUTE.slice(0, FILE_LIST_ROUTE.indexOf(':folderId')) + ROOT_FOLDER_ID;
```
Comment (dkalinin) [inline ui-components/app/components/login-page/index.js:56]:
```suggestion
        window.location.hash = FILE_LIST_ROUTE.replace(':folderId', ROOT_FOLDER_ID);
```

File: ui-components/app/models/list-items/list-item-factory/index.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import FolderItem from '../folder-item';
+import FileItem from '../file-item';
+
+/**
+ * The factory that creates necessary file and folder items.
+ */
+export default class ListItemFactory {
```
Comment (dkalinin) [inline ui-components/app/models/list-items/list-item-factory/index.js:7]: List item is only one context where file/folder is used. Please make this name more abstract.

File: ui-components/app/services/api-service/index.js [hunk 6]
```diff
@@ -88,20 +88,22 @@ export default class ApiService {
...
     case 401:
       return new AuthorizationError('Not authorized.');
     case 404:
       return new NotFoundError('This item does not exist.');
     case 500:
       return new GeneralServerError('Internal server error.');
     default:
-      return new Error('Unknown error');
+      return response.json().then((body) => {
+        return new Error(body);
```
Comment (dkalinin) [inline ui-components/app/services/api-service/index.js:105]: For other errors only string will be returned, not a JSON.

File: ui-components/app/services/fetch-mock/index.js [hunk 7]
```diff
@@ -60,22 +62,51 @@ export default class FetchMock {
...
-          name: 'Documents',
-          itemsNumber: 20,
-          type: 'folder',
-        },
-        {
-          name: 'photo.png',
-          mimeType: 'image',
-          size: 16,
-          type: 'file',
+    fetchMock.get('glob:/folder/*/content', (url) => {
```
[the remark this answers is not in the document]
Comment (bohdan-meltiukhov) [reply ui-components/app/services/fetch-mock/index.js:65]: The express style path does not allow getting route parameters without external tools. Do I still need to change the glob pattern to the express pattern?

File: ui-components/app/state/actions/get-folder-action/index.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    this._folderId = id;
+  }
+
+  /** @inheritdoc */
+  apply(stateManager, apiService) {
+    apiService.getFolder(this._folderId)
+      .then((folder) => {
+        stateManager.mutate(new FolderMutator(folder.folder));
+      });
```
[the remark this answers is not in the document]
Comment (bohdan-meltiukhov) [reply ui-components/app/state/actions/get-folder-action/index.js]: I added it now.

File: ui-components/app/components/breadcrumbs/index.js [hunk 9]
```diff
@@ -68,6 +57,29 @@ export default class Breadcrumbs extends Component {
...
+   * Sets whether the folder data is being loaded or not.
+   *
+   * @param {boolean} isLoading - The flag that shows if the folder data is being loaded.
+   */
+  set isLoading(isLoading) {
+    if (isLoading) {
+      this._folderIcon.style.visibility = 'hidden';
+      this._folderName.innerText = 'Loading...';
+    }
+  }
```
[the remark this answers is not in the document]
Comment (bohdan-meltiukhov) [reply ui-components/app/components/breadcrumbs/index.js:74]: Done. This setter now hides and shows the "Loading..." message.

File: ui-components/app/components/breadcrumbs/index.js [hunk 10]
```diff
@@ -69,7 +71,10 @@ export default class Breadcrumbs extends Component {
   set isLoading(isLoading) {
     if (isLoading) {
       this._folderIcon.style.visibility = 'hidden';
-      this._folderName.innerText = 'Loading...';
+      this._folderName.innerText = '';
```
Comment (dkalinin) [inline ui-components/app/components/breadcrumbs/index.js:74]: - [ ] Let's hide "name" span instead.

File: ui-components/app/services/fetch-mock/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (dkalinin) [reply ui-components/app/services/fetch-mock/index.js:65]: Yes.

**Pull request merged**